aio/tools/examples/create-example.js

```diff
@@ -0,0 +1,137 @@
+
+if (require.main === module) {
+  const options =
+      yargs(process.argv.slice(2))
+          .command(
+              '$0 <name> [source]',
+              [
+                'Create a new <name> example.',
+                '',
+                'If [source] is provided then the relevant files from the CLI project at that path are copied into the example.',
+              ].join('\n'))
+          .strict()
+          .version(false)
+          .argv;
+
+  const exampleName = options.name;
+  const examplePath = path.resolve(EXAMPLES_BASE_PATH, exampleName);
+
+  console.log('Creating new example at', examplePath);
+  createEmptyExample(exampleName, examplePath, {}, {});
+
+  const sourcePath = options.source !== undefined ? path.resolve(options.source) : BASIC_SOURCE_PATH
+  console.log('Copying files from', sourcePath);
+  copyExampleFiles(sourcePath, examplePath);
+
+  console.log(`The new "${exampleName}" example has been created.`);
+  console.log('Now run "yarn boilerplate:add" to set it up for development.');
+}
+
+/**
+ * Create the folder and marker files for the new example.
+ */
+function createEmptyExample(exampleName, examplePath, exampleConfig, stackblitzConfig) {
+  ensureExamplePath(examplePath);
+  writeExampleConfigFile(examplePath, exampleConfig);
+  writeStackBlitzFile(exampleName, examplePath, stackblitzConfig);
+}
+
+/**
+ * Ensure that the new example folder exists.
+ */
+function ensureExamplePath(examplePath) {
+  if (fs.existsSync(examplePath)) {
+    throw new Error(
+        `Unable to create example. The path to the new example already exists: ${examplePath}`);
+  }
+  fs.ensureDirSync(examplePath);
+}
+
+/**
+ * Write the `example-config.json` file to the new example.
+ */
+function writeExampleConfigFile(examplePath, additionalConfig) {
+  fs.writeFileSync(
+      path.resolve(examplePath, EXAMPLE_CONFIG_FILENAME),
+      JSON.stringify(
+          {
+            ...{/* BASIC CONFIG */},
+            ...additionalConfig,
+          },
+          undefined, 2));
+}
+
+/**
+ * Write the `stackblitz.json` file into the new example.
+ */
+function writeStackBlitzFile(exampleName, examplePath, additionalConfig) {
+  const config = {
+    description: titleize(exampleName),
+    files: ['!**/*.d.ts', '!**/*.js', '!**/*.[1,2].*'],
+    file: 'src/app/app.component.ts',
+    tags: [exampleName.split('-')],
+    ...additionalConfig,
+  };
+  fs.writeFileSync(
+      path.resolve(examplePath, STACKBLITZ_CONFIG_FILENAME), JSON.stringify(config, undefined, 2));
+}
+
+/**
+ * Copy all the files from the `sourcePath`, which are not ignored by the `.gitignore` file in the
+ * `EXAMPLES_BASE_PATH`, to the `examplePath`.
+ */
+function copyExampleFiles(sourcePath, examplePath) {
+  const gitignoreFilePath = path.resolve(EXAMPLES_BASE_PATH, '.gitignore');
+  const gitignoreFile = fs.readFileSync(gitignoreFilePath, 'utf8');
+  const gitignore = ignore().add(gitignoreFile);
+  const sourceDir = path.basename(sourcePath);
```

Comment:
It doesn't really matter, but why not use the name of the destination (which the `.gitignore` will actually be applied to) :grin:

```suggestion
  const sourceDir = path.basename(examplePath);
```

Reply:
A philosophical question: are we filtering the paths of the source files or the path of the destination files?

Reply:
As pointed out in https://github.com/angular/angular/pull/39283#discussion_r505657899, probably both :grin:

Reply:
I have changed it to do both!